### DISCUSSION TITLE
Error in WebSocket connection to 1.2.3.4:443: WebSocket Error: INVALID_FRAME_PAYLOAD_DATA (reason: 'utf-8' codec can't decode byte 0xd8 in position 1: invalid continuation byte)

### DISCUSSION THREAD

**Discussion by finalstage** [mitmproxy/mitmproxy#5218] [General]
#### Problem Description
I'm sniffing traffic between an app and their central server using mitmdump (v.8.0.0) in socks5 mode.
Additionally to https requests the app uses a websocket connection to the server.
On some activity in the app it reports that a websocket message was not received (happens only when using mitmproxy).
In the mitmdump output I see the red error message like
**Error in WebSocket connection to 1.2.3.4:443: WebSocket Error: INVALID_FRAME_PAYLOAD_DATA (reason: 'utf-8' codec can't decode byte 0xd8 in position 1: invalid continuation byte)**
The byte position varies, sometimes it was 0xd9.
It is not due to my analysis addon, it happens when I run mitmdump without any scripts too.
It is also not affected by the log/flow level of mitm, the app reports the missing websocket message when going through the proxy.

#### Steps to reproduce the behavior:
It's too complex to set up - but what may I doo to further analyse this problem?

#### System Information
Mitmproxy: 8.0.0
Python:    3.10.4
OpenSSL:   OpenSSL 1.1.1n  15 Mar 2022
Platform:  Windows-10-10.0.19043-SP0

**Comment by Prinzhorn** [comment]
First of here's what the error means: https://github.com/python-hyper/wsproto/blob/7d97af83fa5d375d7645b559523d6e191a6789e2/src/wsproto/frame_protocol.py#L138-L142

>  #: indicates that an endpoint is terminating the connection
>  #: because it has received data within a message that was not
>  #: consistent with the type of the message (e.g., non-UTF-8 [RFC3629]
>  #: data within a text message).

If it's a bug in the proxy, then it's likely in wsproto (I think it's very unlikely that it's in mitmproxy itself). But chances are the app you are using is violating the WebSocket spec.

You can try running `mitmdump --set proxy_debug -vvv` (this will produce _very_ verbose logs) and see if the broken message appears before the Error.

I think you could also chain mitmdump and run the first one (the one in front of the app) in `--raw-tcp` mode. So you can correlate the error in the second mitmdump with the raw tcp traffic in the first and actually see the WebSocket message packets.

**Comment by mhils** [comment]
What @Prinzhorn said - this is a client/server sending invalid WebSocket messages. wsproto always does this form of validation, we cannot easily disable it. Forcing the connection to be treated as plain TCP should at least make the contents visible.

**Comment by finalstage** [comment]
Yes, I expected the problem in a 3rd party module as I couldn't find the error enum in the mitmproxy source. You're right that the problem occurs in wsproto.
I couldn't get proxy chaining to work, so I added lil debug to the wsprotos MessageDecoder::process_frame where the UnicodeDecodeError gets catched (`logging.error(frame.payload)`). I'm really new to python and never had to deal with ws on protocol level..

The payload is protobuf, so shouldn't the message type be binary blob?
W/o decoding the frame in wsproto the exceptions will raise in mitmproxy, don't see how to workaround this.
But if its a client abusing a protocol, shouldn't we still be able to forward the messages?

This is what the log for the payload shows, together wioth the error message of mitmproxy/dumper.

ERROR:root:bytearray(b'\n\xc5\x08\x08\n\x12\xc0\x08\nW\n(7bde6033314e98f42d66248ca65cac2597ec7cb0\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a\x1d\n\x12918118830373757937\x10\x06\x18\t% \xb0kI"\x04\x08\x01\x18\x01\nW\n(f5e3662eb197f10a8017a6c55362abff72fc9921\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a\x1d\n\x12918118830373757937\x10\x11\x18\x01%\x80-\x86G"\x04\x08\x01\x18\x03\nd\n(f716bf543de552cca788486e6cd134fd641d676a\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a,\n m415592fde7821ß4612a472038273773\x10\xbc\x07\x18\x01%\x00\xc0\x81D"\x02\x08\x01\n`\n(56034442688bbe47d1114f17188eb0d60d271806\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a\x1d\n\x12918118830373757937\x10\x07\x18\x04%\x9c\xd1{M"\r\x08\x01\x18\xff\xff\xff\xff\xff\xff\xff\xff\xff\x01\nf\n(df8d6cc37c6b5a5e77be688c2719a8cdc5e5e0c8\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a,\n m415592fde7821ß4612a472038273773\x10\xb6\x01\x18\x01%@N\x85H"\x04\x08\x01\x18\x03\nc\n(175c7e71a1ad1c582eb4ba5ec1c44925e956a48d\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a+\n m415592fde7821ß4612a472038273773\x10.\x18\x01%\x00\x00\x82B"\x02\x08\x01\nc\n(f8d13e520c3b92cf74f81d8b55b563e3008af50b\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a+\n m415592fde7821ß4612a472038273773\x10 \x18\x01%\x00\x00\x82B"\x02\x08\x01\nc\n(0d56e9842cbb39fb60476d49bf2c99ca76e7e345\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a+\n m415592fde7821ß4612a472038273773\x10\x15\x18\x01%\x00\x00\x82B"\x02\x08\x01\nc\n(11488af9689fe613b7b01d6e98269e072d0edc21\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a+\n m415592fde7821ß4612a472038273773\x10?\x18\x01%\x00\x00 B"\x02\x08\x01\nc\n(b21407e453de0c4b22dd19d4534f988ea632e630\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a+\n m415592fde7821ß4612a472038273773\x10*\x18\x01%\x00\x00 B"\x02\x08\x01\nc\n(d407c8751e0f00a6044c2c32a8a7a01b68328490\x12\x06\x08\xa6\xf6\x8b\x92\x06\x1a+\n m415592fde7821ß4612a472038273773\x10)\x18\x01%\x00\x00\x82B"\x02\x08\x01')

Error in WebSocket connection to 52.222.209.37:443: WebSocket Error: INVALID_FRAME_PAYLOAD_DATA (reason: 'utf-8' codec can't decode byte 0xc5 in position 1: invalid continuation byte)

No idea how to proceed, any suggestions?
Thanks very much for your time and sublime work of mitmproxy!

BR
Manuel

**Reply by mhils** [reply]
Proceed where? What's the goal you are trying to accomplish?

**Reply by finalstage** [reply]
It seems like these 'invalid' messages are not arriving the client. They get lost or someway corrupted.
I thought there would be a way to work around this.

**Reply by mhils** [reply]
If you just want to pass this traffic through as-is, take a look at https://docs.mitmproxy.org/dev/howto-ignoredomains/. :)
